### PR TITLE
[TTAHUB-2205] Merge goals backend

### PR DESCRIPTION
**Pull request by thewatermethod**
## Description of change

Merge goals back end, as well as finishing touches on front end

## How to test

Test that you can merge goals, that you are redirected and shown a success message with the merged goals at the top. Please also take a look at the back end script and confirm that the data combinations the merge script performs makes sense

Excluded from this, in #1814 I am parsing the similarity API in order to exclude goals meeting certain conditions and combine them as in the RTR at large. If you see weirdness like what @AdamAdHocTeam describes below, please check to see if you are testing with goals that will be excluded (specifically to avoid weirdness)

**Goals excluded for MVP**: closed curated goals & goals that have no simple linkage back to an older grant

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-2205

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/services/recipient.js [hunk 1]
```diff
@@ -474,33 +474,53 @@ export async function getGoalsByActivityRecipient(
...
       'status',
       'createdAt',
+      'createdVia',
       'goalNumber',
       'previousStatus',
       'onApprovedAR',
       'isRttapa',
       'source',
       'goalTemplateId',
       [sequelize.literal('CASE WHEN COALESCE("Goal"."status",\'\')  = \'\' OR "Goal"."status" = \'Needs Status\' THEN 1 WHEN "Goal"."status" = \'Draft\' THEN 2 WHEN "Goal"."status" = \'Not Started\' THEN 3 WHEN "Goal"."status" = \'In Progress\' THEN 4 WHEN "Goal"."status" = \'Closed\' THEN 5 WHEN "Goal"."status" = \'Suspended\' THEN 6 ELSE 7 END'), 'status_sort'],
```
[the remark this answers is not in the document]
Comment (thewatermethod) [reply src/services/recipient.js:522]: not code from this PR but sure why not

File: src/services/goals.js [hunk 2]
```diff
@@ -2389,3 +2390,277 @@ export async function destroyGoal(goalIds) {
...
+  if (statuses.includes(GOAL_STATUS.CLOSED)) {
+    return GOAL_STATUS.CLOSED;
+  }
+
+  if (statuses.includes(GOAL_STATUS.SUSPENDED)) {
+    return GOAL_STATUS.SUSPENDED;
+  }
+
+  return GOAL_STATUS.NOT_STARTED;
+}
```
[the remark this answers is not in the document]
Comment (thewatermethod) [reply src/services/goals.js]: https://github.com/HHS/Head-Start-TTADP/pull/1799/commits/3200d64f0c4e2cbf991e3fa53ae254c699b662f4

File: src/services/goals.js [hunk 3]
```diff
@@ -2389,3 +2390,277 @@ export async function destroyGoal(goalIds) {
...
+            as: 'objectiveTopics',
+          },
+          {
+            model: ActivityReportObjective,
+            as: 'activityReportObjectives',
+          },
+        ],
+      },
+    ],
+  });
```
[the remark this answers is not in the document]
Comment (thewatermethod) [reply src/services/goals.js]: Good point, addressed this here https://github.com/HHS/Head-Start-TTADP/pull/1799/commits/36745e3bc0d535b1e97a519de5264140196e466a

File: src/services/goals.js [hunk 4]
```diff
@@ -2389,3 +2390,277 @@ export async function destroyGoal(goalIds) {
...
+
+  const finalGoalValues = determineFinalGoalValues(selectedGoals, finalGoal);
+
+  /**
+   * we will need to create a "new" final goal for each grant involved
+   * in this sordid business
+   */
+
+  // unique list of grant IDs
+  const grantIds = uniq(selectedGoals.map((goal) => goal.grantId));
```
[the remark this answers is not in the document]
Comment (thewatermethod) [reply src/services/goals.js]: Addressed this here https://github.com/HHS/Head-Start-TTADP/pull/1799/commits/36745e3bc0d535b1e97a519de5264140196e466a

**Comment by thewatermethod** [comment]
@GarrettEHill Think I addressed all your issues

File: frontend/src/components/GoalCards/GoalDataController.js [hunk 5]
```diff
@@ -72,17 +76,33 @@ function GoalDataController({
...
+          sortBy,
           sortConfig.direction,
           sortConfig.offset,
           goalsPerPage,
           query,
+          mergedGoals || [],
         );
         setData(response);
         setError('');
+        setShouldDisplayMergedSuccess(sortBy === 'mergedGoals' && mergedGoals && mergedGoals.length > 0);
```
Comment (AdamAdHocTeam) [inline frontend/src/components/GoalCards/GoalDataController.js:105]: How does the sortBy relate to a successful merge. Should we track this in its own var?

File: frontend/src/components/GoalCards/__tests__/GoalCards.js [hunk 6]
```diff
@@ -286,6 +286,8 @@ const renderTable = ({ goals, goalsCount, allGoalIds = null }, user, hasActiveGr
             }}
             setGoals={setGoals}
             allGoalIds={allGoalIds || goals.map((g) => g.id)}
+            shouldDisplayMergeSuccess={false}
+            dismissMergeSuccess={jest.fn()}
```
Comment (AdamAdHocTeam) [inline frontend/src/components/GoalCards/__tests__/GoalCards.js:290]: Should we be testing these here or are they tested in another location?

File: frontend/src/components/GoalCards/__tests__/GoalDataController.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+      <UserContext.Provider value={{ user: DEFAULT_USER }}>
+        <Router history={history}>
+          {/* eslint-disable-next-line react/jsx-props-no-spreading */}
+          <GoalDataController {...defaultProps} {...props} />
+        </Router>
+      </UserContext.Provider>,
+    );
+  };
+
+  it('fetches goals in the correct order if specified in history state', async () => {
```
Comment (AdamAdHocTeam) [inline frontend/src/components/GoalCards/__tests__/GoalDataController.js:66]: Should we add tests for the path of history.location.state undefined? Branch coverage.

File: frontend/src/components/GoalCards/__tests__/GoalsCardsHeader.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+      <UserContext.Provider value={{ user: DEFAULT_USER }}>
+        <Router history={history}>
+          {/* eslint-disable-next-line react/jsx-props-no-spreading */}
+          <GoalCardsHeader {...defaultProps} {...props} />
+        </Router>
+      </UserContext.Provider>,
+    );
+  };
+
+  it('displays correct message with merged goals', async () => {
```
Comment (AdamAdHocTeam) [inline frontend/src/components/GoalCards/__tests__/GoalsCardsHeader.js:70]: Do we want to test with alt values for:
shouldDisplayMergeSuccess
dismissMergeSuccess

Also location undefined.

File: frontend/src/pages/RecipientRecord/pages/MergeGoals/index.js [hunk 9]
```diff
@@ -281,6 +285,20 @@ export default function MergeGoals({
...
+  const onSubmit = async (data) => {
+    try {
+      setIsAppLoading(true);
+      const mergedGoals = await mergeGoals(data.selectedGoalIds, data.finalGoalId);
+      const goalIds = mergedGoals.map((g) => g.id);
+      setIsAppLoading(false);
+      history.push(`${backPath}`, { mergedGoals: goalIds });
+    } catch (e) {
+      setValidation('Unable to merge goals');
+      setIsAppLoading(false);
```
Comment (AdamAdHocTeam) [inline frontend/src/pages/RecipientRecord/pages/MergeGoals/index.js:297]: Should this be in finally or will redirect happen if its all good.

File: frontend/src/pages/RecipientRecord/pages/MergeGoals/components/__tests__/FinalGoalCard.js [hunk 10]
```diff
@@ -175,4 +175,34 @@ describe('FinalGoalCard', () => {
...
+        goalStatus: 'Not Started',
+      },
+      {
+        ...baseGoals[1],
+        goalStatus: 'Draft',
+      },
+    ]));
+
+    const label = document.querySelector('.ttahub-final-goal--status .ttahub-final-goal--status-label');
+    expect(label.textContent).toBe('Not started');
```
Comment (AdamAdHocTeam) [inline frontend/src/pages/RecipientRecord/pages/MergeGoals/components/__tests__/FinalGoalCard.js:191]: Should we also be seeing a 'Draft' here?

File: frontend/src/pages/RecipientRecord/pages/MergeGoals/components/__tests__/FinalGoalCard.js [hunk 11]
```diff
@@ -175,4 +175,34 @@ describe('FinalGoalCard', () => {
...
+        goalStatus: 'Draft',
+      },
+      {
+        ...baseGoals[1],
+        goalStatus: 'Draft',
+      },
+    ]));
+
+    const label = document.querySelector('.ttahub-final-goal--status .ttahub-final-goal--status-label');
+    expect(label.textContent).toBe('Draft');
```
Comment (AdamAdHocTeam) [inline frontend/src/pages/RecipientRecord/pages/MergeGoals/components/__tests__/FinalGoalCard.js:206]: Expect two drafts?

File: src/migrations/20231101171425-additional-changes-for-merge-goals.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+            tableName: 'Goals',
+          },
+          key: 'id',
+        },
+      }, { transaction });
+    });
+  },
+  async down(queryInterface) {
+    await queryInterface.sequelize.transaction(async (transaction) => {
+      const sessionSig = __filename;
```
Comment (AdamAdHocTeam) [inline src/migrations/20231101171425-additional-changes-for-merge-goals.js:52]: Can we remove the enum?

File: src/models/goal.js [hunk 13]
```diff
@@ -167,7 +174,7 @@ export default (sequelize, DataTypes) => {
       allowNull: true,
     },
     createdVia: {
```
Comment (AdamAdHocTeam) [inline src/models/goal.js:176]: Do we need column def for 'mapsToParentGoalId' here?

File: src/services/goalServices/getGoalsByActivityRecipient.test.js [hunk 14]
```diff
@@ -793,6 +793,38 @@ describe('Goals by Recipient Test', () => {
...
 
+    it('Retrieves and sorts for merged goals by Recipient', async () => {
+      const { count, goalRows } = await getGoalsByActivityRecipient(recipient3.id, 1, {
+        sortBy: 'mergedGoals', sortDir: 'desc', offset: 0, limit: 20, goalIds: [goalIds[9]],
+      });
+      const countx = count;
+      const goalRowsx = goalRows;
+      expect(countx).toBe(3);
+      expect(goalRowsx.length).toBe(3);
+      expect(goalRowsx[0].id).toBe(goalIds[9]);
```
Comment (AdamAdHocTeam) [inline src/services/goalServices/getGoalsByActivityRecipient.test.js:804]: Should we check second and third position of resulting sorts?

File: src/services/goals.js [hunk 15]
```diff
@@ -2389,3 +2390,314 @@ export async function destroyGoal(goalIds) {
...
+  /**
+   * we will need to create a "new" final goal for each grant involved
+   * in this sordid business
+   */
+
+  const grantsWithReplacements = await Grant.findAll({
+    attributes: ['id', 'status', 'oldGrantId'],
+    where: {
+      [Op.or]: [
+        { id: uniq(selectedGoals.map((goal) => goal.grantId)) },
```
Comment (AdamAdHocTeam) [inline src/services/goals.js:2596]: Prob just rip this out to a var.

File: src/services/goals.js [hunk 16]
```diff
@@ -2389,3 +2390,314 @@ export async function destroyGoal(goalIds) {
...
+  // update associated models
+  // - update AR goals with originalGoalId and goalId with the new goalId
+  // - update goalId on objectives
+  // - copy existing objective, set mapsTo on existingObjective with new objectiveId
+  // - update ARO with originalObjectiveId and objectiveId with the new objectiveId
+  // - goalResources, copy
+  // - goalFieldResponses, copy
+  // - objective files, ETC
+
+  await Promise.all(selectedGoals.map((g) => (
```
Comment (AdamAdHocTeam) [inline src/services/goals.js:2634]: Maybe  comment on this one as it does lots of updating.

File: src/services/goals.js [hunk 17]
```diff
@@ -2389,3 +2390,314 @@ export async function destroyGoal(goalIds) {
...
+  const { dataValues } = objective;
+  const { id, goalId, ...data } = dataValues;
+  const newObjective = await Objective.create({
+    ...data,
+    goalId: parentGoalId,
+    mapsToParentObjectiveId: id,
+  }, { individualHooks: true });
+
+  const updatesToRelatedModels = [];
+
```
Comment (AdamAdHocTeam) [inline src/services/goals.js:2439]: Going to have to remember to update the "merge" code anytime we add goal or objective tables.

File: src/services/goals.test.js [hunk 18]
```diff
@@ -39,6 +54,466 @@ describe('Goals DB service', () => {
...
+    let goalThree;
+
+    let objectiveOneForGoalOne;
+
+    const fileName = faker.system.fileName();
+    const resourceUrl = faker.internet.url();
+
+    beforeEach(async () => {
+      // Recipient.
+      recipient = await Recipient.create({
```
Comment (AdamAdHocTeam) [inline src/services/goals.test.js:118]: Not sure if you want to include a recipient, grant, etc. that should NOT be updated. And just double check.

File: frontend/src/components/GoalCards/GoalDataController.js (hunk 5 above)
Comment (thewatermethod) [reply frontend/src/components/GoalCards/GoalDataController.js:105]: Consult the mockup for details. On a successful merge, the merged goals should display first.

File: frontend/src/components/GoalCards/__tests__/GoalCards.js (hunk 6 above)
Comment (thewatermethod) [reply frontend/src/components/GoalCards/__tests__/GoalCards.js:290]: They are tested in another location

File: src/services/recipient.js [hunk 19]
```diff
@@ -474,33 +474,62 @@ export async function getGoalsByActivityRecipient(
...
   // If we have specified goals only retrieve those else all for recipient.
-  if (goalIds && goalIds.length) {
+  if (sortBy !== 'mergedGoals' && goalIds && goalIds.length) {
     goalWhere = {
       id: goalIds,
       ...goalWhere,
     };
   }
 
+  const sanitizedIds = [
```
Comment (AdamAdHocTeam) [inline src/services/recipient.js:491]: Prob add a comment on this one.

File: src/testUtils.js [hunk 20]
```diff
@@ -263,3 +265,24 @@ export async function destroyGoal(goal) {
...
+
+  const secret = 'secret';
+  const hash = crypto
+    .createHmac('md5', secret)
+    .update(name || n)
+    .digest('hex');
+
+  return GoalTemplate.create({
+    hash,
+    templateName: name || n,
```
Comment (AdamAdHocTeam) [inline src/testUtils.js:285]: var for name || n

File: src/services/goals.test.js [hunk 21]
```diff
@@ -39,6 +54,466 @@ describe('Goals DB service', () => {
...
+      const mergedGoalIds = mergedGoals.map((goal) => goal.id);
+
+      // verify that new goals were created
+      expect(mergedGoalIds).not.toContain(goalOne.id);
+      expect(mergedGoalIds).not.toContain(goalTwo.id);
+      expect(mergedGoalIds).not.toContain(goalThree.id);
+
+      // verify goal & associated data with a length query
+      // Let's get everything and just go through it all
+      const goalsWithData = await Goal.findAll({
```
Comment (AdamAdHocTeam) [inline src/services/goals.test.js:415]: nice

File: frontend/src/components/GoalCards/__tests__/GoalsCardsHeader.js (hunk 8 above)
Comment (thewatermethod) [reply frontend/src/components/GoalCards/__tests__/GoalsCardsHeader.js:70]: This stuff is tested at higher levels, no need to repeat here

File: frontend/src/pages/RecipientRecord/pages/MergeGoals/components/__tests__/FinalGoalCard.js (hunk 10 above)
Comment (thewatermethod) [reply frontend/src/pages/RecipientRecord/pages/MergeGoals/components/__tests__/FinalGoalCard.js:191]: No, not started + draft = not started

File: frontend/src/pages/RecipientRecord/pages/MergeGoals/components/__tests__/FinalGoalCard.js (hunk 11 above)
Comment (thewatermethod) [reply frontend/src/pages/RecipientRecord/pages/MergeGoals/components/__tests__/FinalGoalCard.js:206]: yeah, merging two draft goals yields another draft goal

File: src/migrations/20231101171425-additional-changes-for-merge-goals.js (hunk 12 above)
Comment (thewatermethod) [reply src/migrations/20231101171425-additional-changes-for-merge-goals.js:52]: [not supported](https://www.postgresql.org/docs/current/datatype-enum.html)

File: src/models/goal.js (hunk 13 above)
Comment (thewatermethod) [reply src/models/goal.js:176]: line 111

File: src/services/goalServices/getGoalsByActivityRecipient.test.js (hunk 14 above)
Comment (thewatermethod) [reply src/services/goalServices/getGoalsByActivityRecipient.test.js:804]: sure

File: src/services/goals.js (hunk 15 above)
Comment (thewatermethod) [reply src/services/goals.js:2596]: sure

File: src/services/goals.js (hunk 17 above)
Comment (thewatermethod) [reply src/services/goals.js:2439]: true

File: src/services/goals.js (hunk 16 above)
Comment (thewatermethod) [reply src/services/goals.js:2634]: sure

File: src/services/recipient.js (hunk 19 above)
Comment (thewatermethod) [reply src/services/recipient.js:491]: sure

**Review by AdamAdHocTeam — COMMENTED** [review]
After doing a merge I noticed that the maps to columns are not populating on the goal and objective.

image

File: frontend/src/components/GoalCards/GoalDataController.js (hunk 5 above)
Comment (AdamAdHocTeam) [reply frontend/src/components/GoalCards/GoalDataController.js:105]: Yea I agree but that's not what I'm saying here. We are using a sort order to determine if a sort was successful.

**Review by AdamAdHocTeam — COMMENTED** [review]
For ActivityReportGoals I don't see the old row being updated with the new goalId (59677).

image

File: frontend/src/pages/RecipientRecord/pages/MergeGoals/index.js (hunk 9 above)
Comment (thewatermethod) [reply frontend/src/pages/RecipientRecord/pages/MergeGoals/index.js:297]: Tried it in the finally when I was writing this, didn't work exactly right

File: src/services/goals.test.js (hunk 18 above)
Comment (thewatermethod) [reply src/services/goals.test.js:347]: recipients and grants aren't updated at all, but I added a goal and an objective that weren't updated

**Comment by thewatermethod** [comment]
I believe I addressed your issues & I've added tests to validate

**Review by AdamAdHocTeam — COMMENTED** [review]
Merged Goals: **50556, 54414**

Correct me if I'm wrong but should the old objectives mapTo the new objectives?

image

In this case I would expect to see maps to ids in the 161200 and up. I also see a objective that was not merged not sure if its on purpose (obj 157005).

**Review by AdamAdHocTeam — COMMENTED** [review]
As we discussed, need to handle the case of a goal linked to inactive grant (with no corresponding goals).

**Review by AdamAdHocTeam — COMMENTED** [review]
Merging these two goals ( 46783, 59678) still has a ref to an old objective id in the ARO table.

**Comment by AdamAdHocTeam** [comment]
> As we discussed, need to handle the case of a goal linked to inactive grant (with no corresponding goals).

Lets create a ticket for this just in case we give them the ability to select the goals they want to merge in the future.

**Review by GarrettEHill — APPROVED** [review]

**Review by AdamAdHocTeam — APPROVED** [review]
Overall looks good. 

Some edge cases found but I don't think these will be an issue until we allow the users to select what they want to merge.

I would advise a testing party for this change.

**Comment by thewatermethod** [comment]
Merging - we can resolve additional issues before unlocking the feature flag as we finish up the work on #1798

**Pull request merged**